File: C# Fundamentals/C# OOP Basics/23.Jul-11-NFS/NFS/NFS/Models/Cars/Car.cs
using System;
using System.Text;

public abstract class Car
{
    private string brand;
    private string model;
    private int yearOfProduction;
    private int horsepower;
    private int acceleration;
    private int suspension;
    private int durability;
    public int OverallPerformance => (this.Horsepower / this.Acceleration) + (this.Suspension + this.Durability);
    public int EnginePerformance => (this.Horsepower / this.Acceleration);
    public int SuspensionPerformance => (this.Suspension + this.Durability);
    private string status;
    public abstract void Tune(int tuneIndex, string addOn);
    public void DecreaseDurability(int laps, int length)
    {
        this.Durability -= laps * (length * length);
    }

    protected Car(string brand, string model, int yearOfProduction, int horsepower, int acceleration, int suspension, int durability)
    {
        this.Brand = brand;
        this.Model = model;
        this.YearOfProduction = yearOfProduction;
        this.Horsepower = horsepower;
        this.Acceleration = acceleration;
        this.Suspension = suspension;
        this.Durability = durability;
        this.Status = "Ready";
    }

    public void SetNewStatus(string newStatus)
    {
        this.Status = newStatus;
    }
    public string Status
    {
        get { return this.status; }
        protected set { this.status = value; }
    }
    public string Brand
    {
        get { return this.brand; }
        protected set { this.brand = value; }
    }
    public string Model
    {
        get { return this.model; }
        protected set { this.model = value; }
    }
    public int YearOfProduction
    {
        get { return this.yearOfProduction; }
        protected set { this.yearOfProduction = value; }
    }
    public int Horsepower
    {
        get { return this.horsepower; }
        protected set { this.horsepower = value; }
    }
    public int Acceleration
    {
        get { return this.acceleration; }
        protected set { this.acceleration = value; }
    }
    public int Suspension
    {
        get { return this.suspension; }
        protected set { this.suspension = value; }
    }
    public int Durability
    {
        get { return this.durability; }
        protected set { this.durability = value; }
    }

    public override string ToString()
    {
        StringBuilder sb = new StringBuilder();

        sb.AppendLine($"{this.Brand} {this.Model} {this.YearOfProduction}");
        sb.AppendLine($"{this.Horsepower} HP, 100 m/h in {this.Acceleration} s");
        sb.AppendLine($"{this.Suspension} Suspension force, {this.Durability} Durability");
        

        return sb.ToString().Trim();
    }
}

File: Java/01.Advanced/01.Advanced/11.Classes/Ex/src/t_05_car_salesman/Car.java
package t_05_car_salesman;

public class Car {
    private final String model;
    private final Engine engine;
    private final long weight;
    private final String color;
    private final static String NON_EXISTING = "n/a";

    public Car(String model, Engine engine) {
        this(model, engine, -100, NON_EXISTING);
    }

    public Car(String model, Engine engine, long weight) {
        this(model, engine, weight, NON_EXISTING);
    }

    public Car(String model, Engine engine, String color) {
        this(model, engine, -100, color);
    }

    public Car(String model, Engine engine, long weight, String color) {
        this.model = model;
        this.engine = engine;
        this.weight = weight;
        this.color = color;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder(50);

        sb.append(this.model);
        sb.append(":\n");

        sb.append(this.engine.toString());

        sb.append("  Weight: ");
        sb.append(this.weight == -100 ? NON_EXISTING : this.weight);
        sb.append("\n");

        sb.append("  Color: ");
        sb.append(this.color != null ? this.color : NON_EXISTING);

        return sb.toString().trim();
    }
}
